modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+        methodWriter.dup();
+        methodWriter.invokeConstructor(
+                Type.getType(painlessConstructor.javaConstructor.getDeclaringClass()),
+                Method.getMethod(painlessConstructor.javaConstructor));
+
+        for (ExpressionNode irArgumentNode : irListInitializationNode.getArgumentNodes()) {
+            methodWriter.dup();
+            visit(irArgumentNode, writeScope);
+            methodWriter.invokeMethodCall(irListInitializationNode.getMethod());
+            methodWriter.pop();
+        }
+    }
+
+    @Override
+    public void visitMapInitialization(MapInitializationNode irMapInitializationNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irMapInitializationNode.getLocation());
+
+        PainlessConstructor painlessConstructor = irMapInitializationNode.getConstructor();
+        methodWriter.newInstance(MethodWriter.getType(irMapInitializationNode.getExpressionType()));
+        methodWriter.dup();
+        methodWriter.invokeConstructor(
+                Type.getType(painlessConstructor.javaConstructor.getDeclaringClass()),
+                Method.getMethod(painlessConstructor.javaConstructor));
+
+        for (int index = 0; index < irMapInitializationNode.getArgumentsSize(); ++index) {
+            methodWriter.dup();
+            visit(irMapInitializationNode.getKeyNode(index), writeScope);
+            visit(irMapInitializationNode.getValueNode(index), writeScope);
+            methodWriter.invokeMethodCall(irMapInitializationNode.getMethod());
+            methodWriter.pop();
+        }
+    }
+
+    @Override
+    public void visitNewArray(NewArrayNode irNewArrayNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irNewArrayNode.getLocation());
+
+        List<ExpressionNode> irArgumentNodes = irNewArrayNode.getArgumentNodes();
+        Class<?> expressionType = irNewArrayNode.getExpressionType();
+
+        if (irNewArrayNode.getInitialize()) {
+            methodWriter.push(irNewArrayNode.getArgumentNodes().size());
+            methodWriter.newArray(MethodWriter.getType(expressionType.getComponentType()));
+
+            for (int index = 0; index < irArgumentNodes.size(); ++index) {
+                ExpressionNode irArgumentNode = irArgumentNodes.get(index);
+
+                methodWriter.dup();
+                methodWriter.push(index);
+                visit(irArgumentNode, writeScope);
+                methodWriter.arrayStore(MethodWriter.getType(expressionType.getComponentType()));
+            }
+        } else {
+            for (ExpressionNode irArgumentNode : irArgumentNodes) {
+                visit(irArgumentNode, writeScope);
+            }
+
+            if (irArgumentNodes.size() > 1) {
+                methodWriter.visitMultiANewArrayInsn(MethodWriter.getType(expressionType).getDescriptor(), irArgumentNodes.size());
```

Comment:
pull these out.

Reply:
Used dropped lines here to make it easier to read. Found little benefit in separate variables.